api/models/metadata.go

```diff
@@ -0,0 +1,230 @@
+package models
+
+import (
+	"encoding/json"
+	"fmt"
+	"bytes"
+	"database/sql/driver"
+	"errors"
+)
+
+// Metadata encapsulates key-value metadata. The structure is immutable via its public API and nil-safe for its contract
+// permissive nilability is here to simplify updates and reduce the need for nil handling in extensions - metadata should be updated by over-writing the original object:
+//  oldMd := EmptyMetadata()
+//  newMd := oldMd.With("fooKey",1)
+//  // old MD remains empty
+// Metadata is lenable
+type Metadata map[string]*metadataValue
+
+// metadataValue encapsulates a metadata value in the metadata map,
+// This is stored in its un-parsed JSON format for later (re-) parsing into specific structs or values
+// metadataValue objects are  immutable after JSON load
+type metadataValue struct {
+	val []byte
+}
```

Comment:
a type alias here seems suitable, too (`type metadataValue []byte`) to add methods on top of. not sure if there's plans to add other fields; not very important.

Reply:
i didn't try it but I assumed that I wouldn't be able to unmarshal a value in the same way here (can't assign to the receiver in UnmarshalJSON)  (could just be go n00bz for me though will give that ago).